### PR TITLE
Add full atmosphere restart test

### PR DESCRIPTION
**Pull request by bartgol**
This PR aims at automatizing the AD capability of creating a restart file.

- [x] Fix atm proc group class to correctly expose input fields/groups
- [x] Make the AD create the "RESTART" group
- [x] Create a test that runs twice (much like the io 'restart' test): 1st time for N time step, second time restarts from N/2 timesteps.
- [x] Compare output of the two aforementioned tests.
- [x]  Make restart BFB (requires saving dyn grid state to restart file).

### CODE REVIEW

**Comment by bartgol** [comment]
Update: the test runs, but there are still diffs between the monolithic 2 steps run, and the restarted run. I disabled all physics, and still get diffs, so (part of) the issue is in how the homme interface restarts homme's state. Interestingly, running the output managers right after init during the restarted run, yields a bfb output with the 1st timestep of the monolithic run.

**Comment by bartgol** [comment]
Update: I have made changes in the grid stack, in particular in the SEGrid. It can now store either Continuous Galerkin (CG) or Discontinuous Galerkin dofs (DG) gids. The former corresponds to the old behavior, where corresponding edge dofs across bordering elements have the same gid, while the latter allows (in fact, forces) such dofs to have different gids. This will allow me to use a DG grid to write pio files with native dyn layout, while CG grid fields can be remapped to a PhysicsGLL grid (as before). In practice, all dyn fields will live on a DG grid. The CG grid is only used by the PD remapper, since it makes it possible to form the local map for the remapping.

The current tests now pass, but none of them writes/loads data to/from file with native dynamics layout (the only test that does dyn IO is `dyn_grid_io`, but it remaps to PhysicsGLL grid in the process). The next step is to hack IO classes to allow writing fields in native SEGrid layout.

**Comment by AaronDonahue** [comment]
I think this PR will address #791

**Comment by AaronDonahue** [comment]
I think this will also address #837

**Comment by bartgol** [comment]
@AaronDonahue @tcclevenger @jeff-cohere This PR is probably ready for review. Some tests might still fail on some machines (they all pass on my laptop), but at least we can get a head start on the design review.

I pinged all of you, but I don't expect each of you to review all changes (there's a lot here!). As a guideline, I would like
- @AaronDonahue : review I/O stuff and tests
- @tcclevenger : review homme-related stuff
- @jeff-cohere : review share folder stuff (infrastructure, and abstract classes design)

Of course, if you want to review all of the PR, that'd be great! But I know it's a relatively large one, so I don't expect you to.

Thanks!

File: components/scream/src/share/atm_process/atmosphere_process.hpp
Comment (jeff-cohere) [inline components/scream/src/share/atm_process/atmosphere_process.hpp:166]: Not a big deal at all, but is there a reason you decided to change these method names? To me, `atm_proc.requires_field(f)` is a nice example of  the `noun.relates_to(other_noun)` idiom.

File: components/scream/src/share/atm_process/atmosphere_process_group.hpp
Comment (jeff-cohere) [inline components/scream/src/share/atm_process/atmosphere_process_group.hpp:69]: Any rules about when this can/can't be called?

File: components/scream/src/share/field/field_manager.hpp
Comment (jeff-cohere) [inline components/scream/src/share/field/field_manager.hpp:73]: Maybe "Adds a field to be managed by the FieldManager, that can be used by atmospheric processes." I was confused to read that the "pre-existing" field must not exist already. :-)

File: components/scream/src/share/field/field_property_checks/field_nan_check.hpp
Comment (jeff-cohere) [inline components/scream/src/share/field/field_property_checks/field_nan_check.hpp:119]: Do you think it's worth having an `unflatten(index, rank, dims, indices)` method on the field or layout object with the following signature, in which `dims` and `indices` are of length `rank`, and `i`, `j`, `k`, etc are stored in `indices` as a result of the method?

```
void unflatten(int flattened_index, int rank, int dims[], int indices[])
```

File: components/scream/src/share/field/field_utils.hpp
Comment (jeff-cohere) [inline components/scream/src/share/field/field_utils.hpp:4]: On second thought, maybe `unflatten` could go here as a free function, but we'd have to add the field or layout as an arg.

File: components/scream/src/share/grid/abstract_grid.hpp
Comment (jeff-cohere) [inline components/scream/src/share/grid/abstract_grid.hpp:104]: This function is no longer virtual, so this comment should be removed.

File: components/scream/src/share/grid/abstract_grid.hpp
Comment (jeff-cohere) [inline components/scream/src/share/grid/abstract_grid.hpp:117]: Do you have a one-liner that describes what this checks, and what it does in case the check passes/fails?

Comment (jeff-cohere) [reply components/scream/src/share/grid/abstract_grid.hpp:117]: Also, I don't know where the conversation in the C++ community is regarding the use of exceptions as super-fancy `if` tests, but if we want to programmatically check things and correct them if they're deficient, I would prefer to have this return a boolean that indicates whether the check passed or failed. If, on the other hand, failing this check is always a fatal or uncorrectable error, the exception is more appropriate

File: components/scream/src/share/io/scorpio_input.hpp
Comment (jeff-cohere) [inline components/scream/src/share/io/scorpio_input.hpp:96]: This is a good question! I think the answer is the same as the answer to "is it invalid for this comm to be different from the one used by the grid in the field_mgr?"

File: components/scream/src/share/io/scorpio_input.hpp
Comment (jeff-cohere) [inline components/scream/src/share/io/scorpio_input.hpp:141]: What does it mean for the grid name to be left blank here?

File: components/scream/src/share/io/scorpio_input.cpp
Comment (jeff-cohere) [inline components/scream/src/share/io/scorpio_input.cpp:287]: This smells like something that might go in `field_utils` as well.

File: components/scream/src/share/util/scream_time_stamp.cpp
Comment (jeff-cohere) [inline components/scream/src/share/util/scream_time_stamp.cpp:152]: Indeed.

File: components/scream/src/physics/spa/CMakeLists.txt
Comment (AaronDonahue) [inline components/scream/src/physics/spa/CMakeLists.txt:1]: wow, good catch Luca, this is clearly copy/paste gone wrong.

File: components/scream/src/share/io/tests/io.cpp
Comment (jeff-cohere) [inline components/scream/src/share/io/tests/io.cpp:381]: This might be too much of a pain, but what do you think about `"Field Names"` instead of `"Fields Names"`?

File: components/scream/src/share/io/scorpio_input.cpp
Comment (AaronDonahue) [inline components/scream/src/share/io/scorpio_input.cpp:177]: "the"

File: components/scream/src/share/io/scorpio_output.hpp
Comment (AaronDonahue) [inline components/scream/src/share/io/scorpio_output.hpp:33]: extra whitespace before "ARRAY OF STRINGS"

**Review by jeff-cohere — COMMENTED** [review]
Really good work here, Luca. I've left several comments for your consideration. I stuck mostly to the `src/share` directory but also gave the tests a glance. Let me know if you would like feedback on anything that I haven't commented on.

File: components/scream/src/share/scream_types.hpp
Comment (AaronDonahue) [inline components/scream/src/share/scream_types.hpp:34]: this is great.

File: components/scream/tests/coupled/dynamics_physics/homme_shoc_cld_p3_rrtmgp/homme_shoc_cld_p3_rrtmgp_output.yaml
Comment (AaronDonahue) [inline components/scream/tests/coupled/dynamics_physics/homme_shoc_cld_p3_rrtmgp/homme_shoc_cld_p3_rrtmgp_output.yaml:8]: Does this change need to be made to the other uncoupled and coupled  tests?

**Comment by AaronDonahue** [comment]
This is an epic PR.  My first read through I didn't spot anything that jumped out at me.  I left a few minor comments along the way.  I have a few meetings now, I'm hoping to go through it again after just to make sure since it is so big.

**Review by tcclevenger — COMMENTED** [review]
Nice work, @bartgol. Like Aaron, first look through nothing huge stood out to me. I'll end up looking back through it later to make sure. I found a typo in the homme stuff that I commented on.

File: components/scream/src/share/atm_process/atmosphere_process.hpp
Comment (bartgol) [reply components/scream/src/share/atm_process/atmosphere_process.hpp:166]: I didn't have an equivalent for "internal" fields... So to keep the impl for required, computed, internal somewhat parallel, I switched to this wording. I can change to `has_blah_field(f)`, with blah in `required, computed, internal` maybe?

File: components/scream/src/share/atm_process/atmosphere_process_group.hpp
Comment (bartgol) [reply components/scream/src/share/atm_process/atmosphere_process_group.hpp:69]: Ah, yes, this must be called before any attempt to query the list of internal fields for the group. I will add a note in the comment.

File: components/scream/src/share/field/field_property_checks/field_nan_check.hpp
Comment (bartgol) [reply components/scream/src/share/field/field_property_checks/field_nan_check.hpp:119]: Uhm, it might be useful, but I'm worried about how to use it from device. I am not 100% sure, but I don't think device lambdas can capture host static arrays (such as dims). Though we could declare `int indices[6];` locally, and call `unflatten(idx, 6, dim0, dim1,dim2,dim3,dim4,dim5,indices);`.

If the kokkos layout stored a 1d view for the (physical) extents, then we could probably do it more easily:
```
const auto& layout = f.get_header().get_layout();
const auto extents = layout.extents(); // or dims_view(), or whatever
Kokkos::parallel_for(layout.size(),KOKKOS_LAMBDA(const int idx){
  int indices[6];
  unflatten_indices<6> (idx, extents, indices);
  ...
});
```
Imho, it's a good v1.1 issue.

Comment (bartgol) [reply components/scream/src/share/field/field_property_checks/field_nan_check.hpp:119]: We could offer also an overload where indices are N separate ints, for ease of use. Like:

```
int indices[6];
int i,j,k,l,m,n;
unflatten<6>(idx,extents,indices); // with array
unflatten<6>(idx,extents,i,j,k,l,m,n);
v(i,j,k,l,m,n) = blah; // Concise
v(indices[0], indices[1], indices[2], indices[3], indices[4], indices[5]) = blah; // Lost in parsing...
```

File: components/scream/src/share/field/field_utils.hpp
Comment (bartgol) [reply components/scream/src/share/field/field_utils.hpp:4]: Could also be a (kokkos-friendly) method of `FieldLayout`.

File: components/scream/src/share/grid/abstract_grid.hpp
Comment (bartgol) [reply components/scream/src/share/grid/abstract_grid.hpp:117]: That's a good point. I think I can easily turn them into methods returning bool's. The only place that uses them will always throw if dofs are bad, but returning true/false is definitely more flexible.

File: components/scream/src/share/io/scorpio_input.hpp
Comment (bartgol) [reply components/scream/src/share/io/scorpio_input.hpp:96]: True. I can see how this could be potentially buggy, so maybe I should remove the comm from the inputs. And if one uses that basic ctor (comm+params) and calls `init(field_mgr,grids_mgr)` later, we should check that the stored comm matches the one in the field mgr (and grids mgr)

File: components/scream/src/share/io/scorpio_input.hpp
Comment (bartgol) [reply components/scream/src/share/io/scorpio_input.hpp:141]: This is a poor design choice on my end. It all stems from how the parameter list is shaped, and allowing "simple" syntax for unit tests.

The general use case requires the fields names to be specified like so:
```
Fields:
  Grid Name 1:
     Fields Names: [a, b, c, ...]
  Grid Name 2:
     Fields Names: [a2, b2, ...]
```
But for single-grid tests, it seemed silly to request the user to know/set the grid name, so I allowed the syntax
```
Fields Names: [a, b, c, ...]
```
directly in the root parameter list. Now, the class offers a minimal ctor, where only comm and params are passed. This can be used if a) you don't want to read fields, only some global attributes (e.g., some counters), or if b) you need to read in some info first, before being able to pass in the grids (e.g., spa first reads ncols from nc file, then creates the views, then calls `init`, passing in the newly created views).

So the ctor where the field mgr (and the grid) is passed can call `set_parameters` passing the grid name, and therefore read the sublist `Fields->GridName->Field Names`. OTOH, the other ctors will not pass a grid name, so `set_parameters` will try to look for the `Field Names` entry, and if not found leave `m_fields_names` unset.

Comment (bartgol) [reply components/scream/src/share/io/scorpio_input.hpp:141]: NOTE: I'm not fond of how this ended up. An alternative, is to store the parameter list internally, and query it later for the fields names (if/when the grid name is already available).

File: components/scream/src/share/io/scorpio_input.cpp
Comment (bartgol) [reply components/scream/src/share/io/scorpio_input.cpp:287]: Maybe, yes. The src view is not a field, but we could have a method `void deep_copy(Field<Real>& f, view_1d_type<Real> src);`. I think the `deep_copy` method of `Field` could be implemented in terms of this too...

File: components/scream/src/share/io/tests/io.cpp
Comment (bartgol) [reply components/scream/src/share/io/tests/io.cpp:381]: Not a pain at all. I actually wondered quite a bit what was the correct English formulation between the two. If you say `Field Names` is the right one, I can switch them all.

File: components/scream/tests/coupled/dynamics_physics/homme_shoc_cld_p3_rrtmgp/homme_shoc_cld_p3_rrtmgp_output.yaml
Comment (bartgol) [reply components/scream/tests/coupled/dynamics_physics/homme_shoc_cld_p3_rrtmgp/homme_shoc_cld_p3_rrtmgp_output.yaml:8]: I thought I changed all the yaml files. But I now see I missed some, probably due to the rebase following your tests folder changing PR. I'll fix those.

File: components/scream/src/share/io/scorpio_input.cpp
Comment (bartgol) [reply components/scream/src/share/io/scorpio_input.cpp:287]: OTOH, implementing `Field::deep_copy(src)` in terms of the new util would create a circular dep between `field.hpp` and `field_utils.hpp`... So I might punt on this.

File: components/scream/src/share/field/field_utils.hpp
Comment (bartgol) [reply components/scream/src/share/field/field_utils.hpp:4]: I went for this file, with some help from FieldLayout.

**Comment by bartgol** [comment]
@jeff-cohere @AaronDonahue @tcclevenger I addressed most of your comments. Jeff, I still haven't addressed the `is_required_field` comment. Let me know if there's an option you like better than the current one. I would really like to keep a symmetry between required, computed, and internal.

File: components/scream/src/share/atm_process/atmosphere_process.hpp
Comment (jeff-cohere) [reply components/scream/src/share/atm_process/atmosphere_process.hpp:166]: Aha, makes sense. I like `atm_proc.has_blah_field(f)`, actually. If that's okay with you, let's do it!

**Comment by jeff-cohere** [comment]
> @jeff-cohere @AaronDonahue @tcclevenger I addressed most of your comments. Jeff, I still haven't addressed the `is_required_field` comment. Let me know if there's an option you like better than the current one. I would really like to keep a symmetry between required, computed, and internal.

Looks great! I think `is_{required,computed,internal}_field` is a good solution. Thanks for all the hard work, dude!

**Review by jeff-cohere — APPROVED** [review]

**Review by tcclevenger — APPROVED** [review]

**Comment by bartgol** [comment]
A few last fixes were needed. Things should work now.

**Comment by bartgol** [comment]
There's still one failure on blake. The model restart does not appear to be BFB in release mode. I will investigate.

**Review by AaronDonahue — APPROVED** [review]
I did my second read-through.  Epic PR.  It looks like its failing the AT so we will want to check on that, but otherwise I think this is ready to go.

**Comment by bartgol** [comment]
Update: if Homme is disabled, then the restart test passes on blake (release build). OTOH, the test with Homme fails even with pack size 1 (so it's not pack-related).

**Comment by bartgol** [comment]
I take it back. Pack size 1 seems to pass on blake, hinting at some non BFB issue related to vectorization. This is somewhat odd, considering that pack size 16 is also tested on mappy, where it passes. It might be there's an operation done on packed data during homme's post processing but on unpacked data during restart bootstrap (or viceversa), which only intel happens to expose.

**Comment by bartgol** [comment]
@AaronDonahue I think I fixed the restart test. If you re-approve, we can merge if/when the AT passes.

**Review by AaronDonahue — APPROVED** [review]
Great! Glad you found the issue

**Comment by bartgol** [comment]
Spoke too  soon...But I swear I got it passing on blake at some point. Not sure what happened.

**Comment by bartgol** [comment]
As soon as #1349 is merged, this should pass. @AaronDonahue @jeff-cohere @tcclevenger , if you want to give a final review, there's only a handful of commits since your reviews, so it should be quick (if you limit the review to the last commits).

**Review by tcclevenger — APPROVED** [review]
Looks good to me.

**Review by jeff-cohere — APPROVED** [review]
Looks good, Luca.

**Review by AaronDonahue — APPROVED** [review]
Looks good.  Glad we can finally get this merged.

**Comment by bartgol** [comment]
#1349 went in, so now I'm manually testing on mappy again, to verify everything still works.

**Comment by bartgol** [comment]
I fixed some issues related to hydro-vs-nonhydro mode for Homme. I want to test again on weaver before merging, but it's currently down, so I'll likely pick this up on Monday.

**Comment by AaronDonahue** [comment]
@bartgol , does this PR need to re-approved and tested on mappy?  With #1349 merged it should be ready to go right?

**Review by AaronDonahue — APPROVED** [review]

**Comment by bartgol** [comment]
@AaronDonahue Well, almost. There was still an issue, and tests failed everywhere. I fixed it with the last commit, but it could not be tested cause weaver was offline. Let's give the AT one last round of testing before merging. Meanwhile, you can review the last commit, and approve/comment.

**Comment by bartgol** [comment]
@AaronDonahue Can I merge this?

**Comment by bartgol** [comment]
Retesting, since this PR became stale.

**Comment by AaronDonahue** [comment]
@bartgol , yes.  I suppose once retesting is finished and passes I say go for it.

File: components/scream/src/share/util/scream_time_stamp.cpp
Comment (AaronDonahue) [inline components/scream/src/share/util/scream_time_stamp.cpp:68]: Do we really need an upper bound on the year?  I'm sure we won't be running simulations for a millennium, so maybe this isn't an issue.  But it feels like an unnecessary limitation.  Who knows what the future holds in terms of simulations.

**Review by AaronDonahue — APPROVED** [review]
Just one comment that shouldn't hold this PR going in.

File: components/scream/src/share/util/scream_time_stamp.cpp
Comment (bartgol) [reply components/scream/src/share/util/scream_time_stamp.cpp:68]: Well, I wanted to put a check that caught possibly uninitialized inputs. No really compelling reasons though.

**Pull request merged**
